org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -116,7 +117,39 @@ public void becomeDelegateFor(BranchLabel otherLabel) {
 	this.forwardReferences = mergedForwardReferences;
 	this.forwardReferenceCount = indexInMerge;
 }
-
+private boolean shouldThrow = true;
+protected void trackStackDepth(boolean branch) {
+	/* Control can reach an instruction with a label in two ways: (1) via a branch using that label or (2) by falling through from the previous instruction.
+	   In both cases, we KNOW the stack depth at the instruction from which control flows to the instruction with the label.
+	   Control cannot reach the instruction with a label by falling through if the previous instruction completes abruptly via a goto/return/throw etc
+	   this.codeStream.lastAbruptCompletion == this.position ==> instruction prior to the one with the label is a goto/return/throw
+	*/
+	boolean sourceDepthKnown = branch || this.codeStream.lastAbruptCompletion != this.position;
+	if (this.targetStackDepth == -1) {
+		// First branch to this label || placement of this label
+		if (sourceDepthKnown) {
+			if (this.codeStream.stackDepth < 0) {
```

Comment:
Now we are at it, just checking whether it makes sense to put a check at the decrement parts - yes, there are 193 references for accesses, and this incr/decr wrapping may slow-down - however, we would find the -ve at the time of being assigned. A check here is definitely better than current stage but it still finds the -ve only after it has happened and not at the time of it going -ve. can discuss

Reply:
This is supposed to be extremely rare scenario. When required using conditional watchpoints can be an option rather than change all over the place.